### ISSUE TITLE
Support cygwin git

### ISSUE THREAD

**Issue by rcarmo** [Microsoft/vscode#7998]
- VSCode Version: 1.2.1
- OS Version: Windows 10 Build 14367

Steps to Reproduce:
1. Install Cygwin
2. Edit .vscode/settings.json to set `git.path` to point to `"\\cygwin64\\bin\\git.exe`
3. VS Code throws up an error message of `ENOENT: no such file or directory, lstat 'C:\cygdrive'`, as described [here](http://stackoverflow.com/questions/36508177/visual-studio-code-cannot-detect-cygwin-git-exe-path)

This is, of course, because Cygwin namespaces the filesystem differently. However, doing this (which usually fixes things in other tools):

```
        "git.path": "\\cygwin64\\bin\\bash.exe git",
```

...then results in further confusion as Code apparently does not have a way to correctly set the environment for the git process, which then defaults to the wrong PATH:

```
git fetch
git show HEAD:.vscode/settings.json
git show HEAD:.vscode/settings.json
Could not create directory '/c/cygwin64/home/USERNAME/.ssh'.
ssh_askpass: exec(/usr/local/bin/win-ssh-askpass.exe): No such file or directory
Host key verification failed.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
git fetch
Could not create directory '/c/cygwin64/home/USERNAME/.ssh'.
ssh_askpass: exec(/usr/local/bin/win-ssh-askpass.exe): No such file or directory
Host key verification failed.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Trying to set PATH, HOME, etc. inline in the `bash` command line above fails for some reason I cannot yet fathom, but I would suggest a configuration setting `git.env` be created so that the environment external Git tools can be pre-configured (this would also benefit the Linux and OSX ports).

As to why this is necessary and why I can't use other Git binaries on Windows, the reason is simple: 

Cygwin does Git over SSH "right", with completely identical setup in terms of SSH keys, options, CR/LF encoding, etc. Current Git for Windows simply does not work for my use case (I've tried), and, obviously, the Linux Subsystem doesn't help with this either.

**Comment by joaomoreno** [comment]
Duplicate of #7361. We don't support cygwin git, sorry.

We've been using Git for Windows successfully in the team, complete with SSH keys, options, line ending encodings et al. What is the actual issue you get with it?

**Comment by rcarmo** [comment]
Hmmm. So this is a dupe of a ticket that's closed because we've "just never really looked into it"?

**Comment by joaomoreno** [comment]
If #7361 were open we would close this one the same way. It is a duplicate, in any case.

Apologies for coming out harsh.

Let me expand a bit on *just never really looked into it*. Code already needs to support whichever version of git you have installed in whichever platform you are using it. It is a non-trivial task that gets another complexity fold as soon as we add git *flavours*. We've found that Git for Windows handles the great majority of the tasks users need and I've personally have yet to find a reason that would make me use cygwin instead.

Of course, I can be wrong and I'm definitely interested in someone proving me wrong. That's why I asked about the actual issue you face. If it is really something Git for Windows does not support, of course we'll reconsider our position towards cygwin, and we'll work towards that.

**Comment by rcarmo** [comment]
It's all about integration. When you use Cygwin and its userland (bash, CLI tools, SSH, etc.), you don't need an extra terminal emulator like PuTTY, you can use standard Git plugins that rely on Perl/Python etc. seamlessly, and you don't need to set up a copy of your SSH keys anyplace else. 

Right now, the current (stunted and incomplete) UNIX userland clone that ships with Git for Windows doesn't provide me those affordances. Also, I can't set GIT_AUTHOR_* and GIT_COMMITTER_* on a per-repo basis using Git for Windows - it's just a surreal amount of pain to deal with that kind of thing inside it, when it all works perfectly in Cygwin.

With an environment setting I might not only actually get Cygwin Git to work (I suspect it's just a matter of setting the right HOME and a few more variables) but also have GIT_* settings set on a per-project basis for me inside VS Code. And this last bit will benefit every other platform, too.

**Comment by joaomoreno** [comment]
Alright, let's keep this open for tracking the support. A PR would be much appreciated as we'll likely not have cycles to look into it any time soon.

Maybe next time I'm in the Lisbon office, I'll ping you and we'll get it working in an afternoon. :wink:

**Comment by Rigg802** [comment]
I created a symlink and it worked for me:
http://stackoverflow.com/questions/36508177/visual-studio-code-cannot-detect-cygwin-git-exe-path/37980938#37980938

**Comment by joaomoreno** [comment]
@rcarmo Did you try this?

**Comment by rcarmo** [comment]
Not yet, kind of tied up with FY16 still. Will give it a go next weekend and see if it works on my box and a fresh install (never underestimate the power of upgrades where it regards breaking things).

**Comment by Rigg802** [comment]
someone had a similar issue here too and it worked: https://github.com/Microsoft/vscode/issues/1982

**Comment by rcarmo** [comment]
I'm going to have a go at this (both reproducing and understanding _exactly_ why the symlink works, and if it works with my git plugins) within the next few days and will report back.

**Comment by rcarmo** [comment]
Brief status update: I'm still having trouble with SSH/askpass at this point. Trying to debug.

**Comment by rogersachan** [comment]
@rcarmo Would Bash for WIndows be an option?

**Comment by rcarmo** [comment]
I've been using it alongside Cygwin, and the answer is no. Unless I go full X-Windows and run VS Code inside LXSS, I can't get the degree of integration I need (plus the LXSS console is still not as nice as mintty, but that's beside the point here)

I made some headway recently using a little executable shim to wrap Cygwin's git, but I'm still missing something regarding PATH and environment variables. I still think being able to set environment variables prior to Git execution is the way around this, but time hasn't been on my side recently.

**Comment by jastorgalj** [comment]
Solution for me (Windows 10 - git from cygwin)
create dir - mkdir c:\cygdrive
open a cmd and goto  - cd c:\cygdrive
créate link to c:\ - mklink /j "c" c:\    
and ready..

**Comment by rcarmo** [comment]
That doesn't work for me for some reason - I get errors trying to find my home directory for SSH keys (Cygwin apparently gets confused and does stuff like `/c/cygwin64/home/username/.ssh`

**Comment by rcarmo** [comment]
I had partial success today by setting a couple of links and tweaking environment variables and managed to have Cygwin git do local commits, but there were two issues:

- the sync icon turned into a cloud+arrow icon (which means VSCode wasn't able to completely figure out the output from the Cygwin git binary). Pushing to the remote repo worked, but...
- ssh-ask-pass did not work correctly, and did not respect the environment variables set by ssh-agent (so I got an SSH password prompt on every git operation)

The second point essentially circles back to one of my earlier points - without being able to set environment variables for git within VS Code, this will likely always be broken in some way -- some people might get it to work with passwordless keys, but it's not a definitive solution.

Also (and this is just an aside), getting this to partially work completely broke "regular" Windows Git (which uses a variant of Cygwin):

```
git pull
      0 [main] ssh 9472 C:\Program Files\Git\usr\bin\ssh.exe: *** fatal error - add_item ("\??\C:\Program Files\Git", "/", ...) failed, errno 1
Stack trace:
Frame        Function    Args
000FFFF9BB0  0018007164E (00180273652, 00180224DD9, 000FFFF9BB0, 000FFFF8B30)
000FFFF9BB0  00180046669 (000FFFFABE0, 000FFFF9BB0, 1D21574853C4D14, 00000000000)
000FFFF9BB0  001800466A2 (000FFFF9BB0, 00000000001, 000FFFF9BB0, 505C3A435C3F3F5C)
000FFFF9BB0  001800DF22B (00000000000, 000FFFFCE00, 001800CEE68, 1D215748E2D2E24)
```

**Comment by joaomoreno** [comment]
The _cloud+arrow_ icon simply means the current branch you have checked out doesn't have an upstream branch configured. Once it does, it should show the sync icon. This must've been a coincidence that you checked out such a branch.

So... would it work if you set a few env variables for when Code calls git?

**Comment by rcarmo** [comment]
@joaomoreno: I believe so, yes. The experiment I did with an executable wrapper a while back was partially successful.

Maybe something like this would be feasible:

```
{
    "git.path": "..."
    "git.env": {
        "foo": "bar"
        ...
    }
}
```

**Comment by joaomoreno** [comment]
And are the variables predictable for each system? Meaning, could it simply be a `git.cygwin: true` setting?

**Comment by rcarmo** [comment]
Nope, I have to set _at least_ these for the SSH agent:

```
SSH_ASKPASS='C:\cygwin64\usr\local\bin\win-ssh-askpass.exe'
SSH_AUTH_SOCK=/tmp/ssh-59iOCz4fyxRZ/agent.12660
```

I can pin down SSH_AUTH_SOCK, but that's usually random, and I will need extra variables for some Git plugins and environments.

**Comment by rcarmo** [comment]
Also, regarding the cloud+arrow bit, there's definitely something amiss here. I've tried Windows Git and Cygwin git on the same repo, same branch, no changes, and gotten:

- cloud+arrow for Cygwin git
- sync wheel for Windows git

Might be a version issue (Cygwin git is 2.8.3)

**Comment by slepkaviba** [comment]
Is there any plan to add cygwin support? It is the only thing missing to make VSC perfect. (IntelliJ has no issue with their git, and I dont like having git-bash just to make VSC work...)

**Comment by escalonn** [comment]
Warning for those using the kluges in the comments here - I was able to work with VSC okay (seemingly), but when I added VS to the mix, it wanted to change all permissions in opened folder `a+x`, while VSC then wanted to change them all `a-x`, so both programs could not see a clean working tree at the same time. Was not able to diagnose the issue further but that was enough for me to say "OK I'll use what is officially supported".

**Comment by rcarmo** [comment]
FWIW, it now sort of works with git inside the Linux Subsystem (aka Bash for Windows), although I keep having trouble with SSH auth.

**Comment by texasbruce** [comment]
Still broken as of today, for cygwin and msys

The workaround is @Radon222 's answer. VSC is trying to read the repo path from \home (msys) or \cygdrive (cygwin) and obviously it does not exist in windows. It should be a simple path conversion?

**Comment by rcarmo** [comment]
Well, no, but “proper” use of Git in many situations requires SSH key auth, and I, for one, vastly prefer it to auth helpers and HTTPS. 

(I’m still using the built-in terminal to run bash and commit via the CLI, but in WSL now.)

**Comment by joaomoreno** [comment]
Sorry, guys, since this is just not a priority for us, I'm making the tough call to close this as out of scope, given that:

- Git for Windows is supported
- WSL support is tracked in #9502

**Comment by DraconPern** [comment]
For anyone that saw this thread doing a search on how to use Cygwin git and vscode, my solution was to install Git for Windows, and then use mklink to junction link the Cygwin .ssh directory to %USERNAME%/.ssh to share config and ssh key.  Then both git will have access to the same ssh key and ssh config for multiple key usage.

**Comment by fromSpace** [comment]
Hello All,

Linked below is a gist that wraps the Cygwin git.exe and converts the Linux paths which cause the error: "error while loading shared libraries: ?: cannot open shared object file: No such file or directory" to Windows paths using the cygpath executable.

https://gist.github.com/nickbudi/4b489f50086db805ea0f3864aa93a9f8

**Comment by nukata** [comment]
I think `cyg-git` might be able to solve the problem.  It provides a "Git for Windows" virtually.  You only need Cygwin and its packages.

https://github.com/nukata/cyg-git

**Comment by Omir85** [comment]
My 2 cents on this (I know, closed and aging) issue: I have added one environmental variable to include my cygwin bin folder to Windows path, and the warning message was removed after restarting VSCode.

**Comment by AnrDaemon** [comment]
Anybody's up for writing a small patch that would call `cygpath -alm` on the results of a few calls VS Code makes to Git that produce a full path as the result?
It's not really a rocket science for somebody familiar with VS Code internals.

**Comment by texasbruce** [comment]
> 
> 
> Anybody's up for writing a small patch that would call `cygpath -alm` on the results of a few calls VS Code makes to Git that produce a full path as the result?
> It's not really a rocket science for somebody familiar with VS Code internals.

I believe this is not an issue with cygpath. cygpath can produce correct windows path.

My assumption is vscode is using the wrong root path while parsing. I might be wrong though, since I didn't check the source code.

**Comment by AnrDaemon** [comment]
This is an issue between VS Code (which relies on git itself too much) and Cygwin git (which uses Cygwin directory layout, which follows POSIX FHS).
To make VS Code understand Cygwin paths, it should either be told to expect them, or use cygpath to convert them to native form.

**Comment by texasbruce** [comment]
> 
> 
> This is an issue between VS Code (which relies on git itself too much) and Cygwin git (which uses Cygwin directory layout, which follows POSIX FHS).
> To make VS Code understand Cygwin paths, it should either be told to expect them, or use cygpath to convert them to native form.

You can create a wrapper for that in Windows batch file, which calls cygpath to output windows path, and VSCode does not have to understand cygwin path, but it still does not work perfectly. The initial Git init works, but the file listing and diff do not work.

I briefly checked the VSCode before. When calling Git, it does not always use absolute path returned from Git command. It parses URI with a root path returned from other function, which does not seem to work properly with non official git binary. You can check that if you like.

**Comment by AnrDaemon** [comment]
Unfortunately, VS Code tries hard to find specifically "git.EXE", even if told to use "git.sh" wrapper (which has appropriate shell association defined).
I've had a small success with "git.cmd" wrapped around "git.sh", trying that route now.

**Comment by AnrDaemon** [comment]
However, a more native approach would be preferred.
Cygwin, despite its interoperability issues on filesystem level, has way better integration with Windows itself. In fact, better to the level that WSL could never achieve. Ever. Due to its nature.

**Comment by AnrDaemon** [comment]
> You can create a wrapper for that in Windows batch file, which calls cygpath to output windows path

Now, I recalled why did I switch from CMD to TakeCommand/Console.
Insane, unusable quoting rules of CMD, compounded by inability to control arbitrary number of batch file parameters in an efficient way, are basically a showstopper to any effective scripting.

**Comment by igordcard** [comment]
Just adding myself here as another person who would really like to continue to use cygwin... + vscode.

**Comment by lalawue** [comment]
> I think `cyg-git` might be able to solve the problem. It provides a "Git for Windows" virtually. You only need Cygwin and its packages.
> 
> https://github.com/nukata/cyg-git

it works for me under Win 7

**Comment by dmak** [comment]
Just tried the script from [Cygwin Git + VS Code compatibility](https://gist.github.com/nickbudi/4b489f50086db805ea0f3864aa93a9f8) and it didn't work with the following error message in Git output:
```
> git ls-files --stage -- D:\Workspace\test\platformio.ini
fatal: D:\Workspace\test\platformio.ini: 'D:\Workspace\test\platformio.ini' is outside repository at '/cygdrive/d/Workspace/test'
```
@nukata
Many thanks for your contribution, it works. Just one minor remark: what about re-writing the script on bash? bash is for sure present on all installations and most are familiar with it. I am not sure if I did this rewriting correctly as I couldn't test the script thoroughly for the case when `wrap_output=yes`, but here it goes:
```
#!/bin/bash

wrap_output=""
argv=("$@")

for ((i = 0; i < $#; i++))
do
	arg="${argv[$i]}"
	if [ $i = 0 -a "$arg" = 'rev-parse' ]
	then
		wrap_output=yes
	elif [[ -n "$arg" && "$arg" != -* ]]
	then
		argv[i]=`cygpath -u "$arg"`
	fi
done

if [ $wrap_output ]
then
	if out=`git "${argv[@]}"`
	then
		if [ -n "$out" ]
		then
			if [[ $out = -* ]]
			then
				echo "$out"
			else
				cygpath -w "$out"
			fi
		fi
	else
		x=$?
		echo "$out"
		exit $x
	fi
else
	exec git "${argv[@]}"
fi
```

**Comment by VincencPodobnik** [comment]
@dmak Thank you! Great work, just one tiny adjustment. Either initially prefix the PATH with `/usr/bin` or change the `git` execs to `/usr/bin/git` or you'll end up with an infinite loop within the script.

**Comment by DanKaplanSES** [comment]
@joaomoreno Clarification: if someone were to submit a PR to fix this, would it be rejected?

I'm asking because if you would conditionally accept it, technically, shouldn't this issue remain open?

**Comment by danfly09** [comment]
But why is it complicated to have git integration with VSCode? 
In my particular case, I don't think I have a problem with the path. It seems that VSCode can detect I have an installation of git and that's because I added to PATH my cygwin bin folder, and then I can execute git commands from CMD and PowerShell without issues. 

However, the editor fails to track the repositories completely.

**Comment by AnrDaemon** [comment]
This is because git reporting paths unexpected to VS Code.
Plus it is reporting them inconsistently. Sometimes relative, sometimes absolute, from the same command.
This is not a simple issue, unfortunately.

**Comment by danfly09** [comment]
```
// Native sandbox environment
if (typeof globals.vscode !== 'undefined' && typeof globals.vscode.process !== 'undefined') {
	const sandboxProcess: INodeProcess = globals.vscode.process;
	safeProcess = {
		get platform() { return sandboxProcess.platform; },
		get arch() { return sandboxProcess.arch; },
		get env() { return sandboxProcess.env; },
		cwd() { return sandboxProcess.cwd(); }
	};
}
```
I was wondering if an additional check can be made here so if sandboxProcess.platform is win32, then check for a Cygwin installation, this can be checked by executing the command, 
`cygcheck --version`
Cygwin bin folder needs to be added to the Path variable. A user configuration variable can also do the work of checking whether VSCode should run natively or under a Cygwin session.

In that scenario sandboxProcess.platform should be linux and not win32, and I believe the compatibility for a real linux OS should work too. 

Another option might be a container, but I have no experiencie with VScode containers or extensions, but the idea is there, to be implemented.

**Comment by AnrDaemon** [comment]
You can have multiple Cygwin environments installed on the same PC. And none of them in the $PATH.
I've long since abandoned the hope and installed Git-for-Windows (and told VS Code to use specifically that one).

**Comment by ViPod5** [comment]
I've been using this https://github.com/nukata/cyg-git with success.

**Comment by maksim-pinguin** [comment]
as noted in #148720 i believe supporting the configured integrated terminal is the way to go. extensions like gitlense do this e.g.

**Comment by AnrDaemon** [comment]
This is not enough.
Cygwin paths are inherently different from Windows paths, and VS Code do not understand that distinction.

**Comment by ohhmm** [comment]
Vscode could have Git auto-detection for Windows from different sources:

- Program Files\Git
- Visual Studio Git
- Cygwin git could be auto-detected by its exe file location

It may be hardcoded preference or highest version preferred.

**Comment by AnrDaemon** [comment]
> Vscode could have Git auto-detection for Windows from different sources:
> 
> * Program Files\Git
> * Visual Studio Git
> * Cygwin git could be auto-detected by its exe file location
> 
> It may be hardcoded preference or highest version preferred.

Yet again - this is not about git.exe location, it is about Cygwin paths support.

**Comment by ohhmm** [comment]
Yes, I got it @AnrDaemon 
Could you try to run Cygwin's git from cmd by its Windows path, please?

**Comment by AnrDaemon** [comment]
Yes, you could. That's not the problem.
The problem starts when Cygwin Git return Cygwin paths which VS Code don't know how to interpret.
Or when VS Code provide Windows paths that means nothing to Cygwin Git.

It requires some extra care inside VS Code itself, and even then not fully fool proof, as other extensions may run Git directly by themselves.

**Comment by DanKaplanSES** [comment]
@AnrDaemon Yep. it's not just the difference in slashes, absolute paths and cygwin start with `/cygdrive/<drive letter>` and in Windows, start with `<drive letter>:\\`. That last paragraph is especially tricky.

Here's my workaround, and I didn't come up with it on my own. In fact, I may be summarizing something already in this thread.

1. I added this line to `settings.json`: `  "git.path": "C:\\cygwin64\\home\\<user>\\vscode_git_compatibility\\cygpath-git-vscode.bat",`.
2. `$ cat cygpath-git-vscode.bat`:
   ```bat

   @echo off

   REM wrapper to convert linux paths to windows
   REM so vscode git integration will work with cygwin
   REM "git.path"="C:\\this\\file.bat" in settings.json

   setlocal
   set PATH=C:\cygwin64\bin;%PATH%

   if "%1" equ "rev-parse" goto rev_parse
   git %*
   goto :eof
   :rev_parse
   for /f %%1 in ('git %*') do cygpath -w %%1
   ```
3. I don't know why, and I don't know if it's actually used, but I have another file in that same directory named `cygpath-git-editor.sh`. `$ cat cygpath-git-editor.sh`:

   ```bash
   #!/bin/bash
   
   # wrapper to convert linux paths to windows
   # so vscode will work as a git editor with cygwin
   # editor="/home/this/file.sh" in .gitconfig

   # extract last argument (the file path)
   for last; do true; done

   # get all the initial command arguments
   all="${@:1:$(($#-1))}"

   # launch editor with windows path
   code $all $(cygpath -w $last)
   ```

I'm kind of in a rush so I don't have the time to verify if these are the only steps, but it is what I currently use. I hope this can be of help to someone.

**Comment by AnrDaemon** [comment]
The workaround is to install Git-For-Windows in a separate directory and point VS Code to it.

**Comment by DanKaplanSES** [comment]
> The workaround is to install Git-For-Windows in a separate directory and point VS Code to it.

Are providing an alternative workaround or explaining mine?

**Comment by AnrDaemon** [comment]
Restating the obvious.

**Comment by DanKaplanSES** [comment]
> Restating the obvious.

Not obvious enough, it seems. [It looks like you said this back in 2022](https://github.com/microsoft/vscode/issues/7998#issuecomment-1100655030), but I hadn't noticed until now. 

If you use the git-for-windows work around, can you still use cygwin's `git` in the same project, or do they step on each other's toes? I use the git command line 95% of the time; I only need VS Code to understand enough to mark my files as modified.

**Comment by AnrDaemon** [comment]
I do.

Cygwin is my go-to *NIX tools, I only install native ports/equivalent tools when one of the following statements is true:

1. The application provides exclusively network functionality or otherwise not restricted by path mangling (webservers, tools that talk to STDIO exclusively). Native tools usually more performant because they are not using POSIX API conversion layer.
2. An external native application wants to pass paths to the tool, and I can not control what exactly is passed. (VS Code talking to Git, Subversion, Meld etc.)

**Comment by edemaine** [comment]
For what it's worth, I tried to make a more robust shim solution, inspired by [cyg-git](https://github.com/nukata/cyg-git) which I first learned about on this thread.

[Cygshim](https://github.com/edemaine/cygshim) should be easier to install (single prebuilt binary, then update PATH) and should handle edge cases better. In addition to Git, it provides LaTeX shims for LaTeX Workshop. If you have ideas for further improvements or extensions (e.g. other commands that need shims), let me know (on cygshim repo not here).
